[doc] Completely reorganize installation instructions.

jadephilipoom opened the pull request:
Resolves #11217 
Related to #11581 

This PR basically reorganizes the material in all our getting-started guides to match the outline in the issue linked above. The diff is large, but mostly it's not new material, just moving around and updating the things we already have. In summary:
- The "Getting Started" page is now the main and only landing page for OpenTitan setup. It includes installation instructions that can be followed from zero to `hello_world` (reworked material from the `install_instructions.md`, which has been deleted)
- Users without Linux are now clearly instructed that they can only use the Docker containers
- I reorganized the files so that the "getting started" material is no longer under "user guides", to avoid distracting experienced users with new-starter material and vice versa
- I emphasized which installation instructions are optional and explained when one might need them
- The outdated "Quickstart" page is deleted

This PR might be convenient to review commit-by-commit, because of file moves.

timothytrippel commented on doc/getting_started/_index.md:
```diff
@@ -0,0 +1,198 @@
+* *If you want to contribute SystemVerilog code upstream to OpenTitan*, follow step 6a to install Verible.
+* *If you want to interact with the chip over JTAG or use GDB for OpenTitan software*, follow step 6b to install OpenOCD.
+
+It also may make sense to stick with the basic setup and come back to these steps if you find you need them later on.
```
nit:
```suggestion
It also may make sense to stick with the basic setup and come back to these steps if you find you need them later.
```

timothytrippel reviewed (commented):
Thanks @jadephilipoom, I have only partially reviewed (I will review the rest over the next day or so) and left some comments, but this is already a major improvement!

jadephilipoom commented on doc/getting_started/_index.md:
```diff
@@ -0,0 +1,198 @@
+* **Option 1 (default, recommended for new users):** [Verilator guide]({{< relref "setup_verilator.md" >}}), or
+* Option 2 (FPGA setup): [FPGA guide]({{< relref "setup_fpga.md" >}}), or
+* Option 3 (design verification setup): [DV guide]({{< relref "setup_dv.md" >}})
+* Option 4 (formal verification setup): [Formal Verification guide]({{< relref "setup_formal.md" >}})
```
Optional steps makes sense, yep! For the other optional steps, I have a little blurb saying "if X, then follow step Y to do Z". For the formal verification flow, what should X be? "If you want to set up formal verification" is the best I can come up with but it would be better to have something a little more informative :slightly_smiling_face:

tjaychen commented on doc/getting_started/setup_verilator.md:
```diff
@@ -128,12 +167,12 @@ $ echo 'h09 l31' > gpio0-write  # Pull the pin 9 high, and pin 31 low.
 
 
-## Connect with OpenOCD to the JTAG port and use GDB
+## Connect with OpenOCD to the JTAG port and use GDB (optional)
```
o does this mean the fpga documentation never had the gdb instructions to begin with?

timothytrippel replied:
The FPGA and Verilator docs both did (bottom of https://docs.opentitan.org/doc/ug/getting_started_fpga/ and https://docs.opentitan.org/doc/ug/getting_started_verilator/) but this just makes it clear this step is "optional" I believe.

timothytrippel commented on util/container/README.md:
```diff
@@ -8,6 +8,7 @@ software development tools for OpenTitan, as listed in the
 
 Skip this step if planning to use the pre-built container. To build in local
 mode:
+<!-- TODO: Where can one find the pre-built container mentioned above? --->
```
I think these use to be packaged in our [releases](https://github.com/lowRISC/opentitan/releases), which are rather stale at this point.

Additionally, I believe our CI builds it too. @milesdai can confirm? 

Maybe this is something CI could also drop in the public GCS bucket along with prebuilt FPGA bitstreams? @arunthomas

milesdai replied:
I couldn't quite find it in the releases when I downloaded it. The CI does in fact build 3 different containers (including the developer utility container referenced here), but the final image is not uploaded anywhere or made available as an artifact. As far as I can tell, you need to build the docker image yourself.

It's one of the things I want to look into for the CI later on, but for now, we may want to just omit the text about the pre-built container since the state of the Dockerfile is not entirely clear at the moment.

timothytrippel replied:
Ah got it, thanks @milesdai !

So maybe just stick with the text that says how to build it.

timothytrippel commented on sw/host/spiflash/README.md:
```diff
@@ -42,7 +42,7 @@ Where ${DEVICE} is one of 'sim_verilator' or 'fpga_nexysvideo'
 ## Run the tool with the Verilator Simulation
 
-Run Verilator with test_rom as described in the [verilator]({{< relref "doc/ug/getting_started_verilator" >}}) getting started guide.
+Run Verilator with test_rom as described in the [verilator]({{< relref "doc/getting_started/setup_verilator" >}}) getting started guide.
```
```suggestion
Run Verilator with test ROM as described in the [verilator]({{< relref "doc/getting_started/setup_verilator" >}}) getting started guide.
```

timothytrippel commented on sw/host/spiflash/README.md:
```diff
@@ -26,7 +26,7 @@ $ ninja -C build-out sw/host/spiflash/spiflash_export
 
 ## Setup instructions for Verilator and FPGA
-Please refer to [verilator]({{< relref "doc/ug/getting_started_verilator" >}}) and [fpga]({{< relref "doc/ug/getting_started_fpga" >}}) docs for more information.
+Please refer to [verilator]({{< relref "doc/getting_started/setup_verilator" >}}) and [fpga]({{< relref "doc/getting_started/setup_fpga" >}}) docs for more information.
```
```suggestion
Please refer to [Verilator]({{< relref "doc/getting_started/setup_verilator" >}}) and [FPGA]({{< relref "doc/getting_started/setup_fpga" >}}) docs for more information.
```

timothytrippel commented on doc/getting_started/setup_verilator.md:
```diff
@@ -21,16 +55,13 @@ First the simulation needs to built itself.
-
 
 Then we need to build software to run on the simulated system.
 There are 3 memory types: ROM, RAM and Flash.
```
Technically there are 4 memories: ROM, SRAM, Flash, and OTP. Images are only built for ROM, Flash, and OTP. SRAM gets populated at runtime.

timothytrippel commented on doc/getting_started/setup_verilator.md:
```diff
@@ -54,11 +85,16 @@ $ build/lowrisc_dv_chip_verilator_sim_0.1/sim-verilator/Vchip_sim_tb \
 To stop the simulation press CTRL-c.
 
-## Interact with the simulated UART
+If you want to run a program other than `hello_world`, you'll change the second `--meminit` argument to point to a different `.scr.vmem` file under `build-bin`.
```
```suggestion
If you want to run a program other than `hello_world`, you'll change the second `--meminit` argument to point to a different `.64.scr.vmem` file under `build-bin`.
```

timothytrippel commented on doc/getting_started/setup_dv.md:
```diff
@@ -1,10 +1,12 @@
+title: "Design Verification Setup"
 ---
 
+_Before following this guide, make sure you've followed the [dependency installation and software build instructions]({{< relref "getting_started" >}})._
```
Since this message is the same across many docs, maybe it would be good idea to create a [shortcode layout, like we do for tool versions](https://github.com/lowRISC/opentitan/blob/master/site/docs/layouts/shortcodes/tool_version.html) (see the Verilator [example](https://github.com/lowRISC/opentitan/blob/356c0543eec55b5b583592329ffa787612d06830/doc/ug/install_instructions/index.md?plain=1#L67))? That way you can keep these messages in sync.

jadephilipoom replied:
They're not quite all the same -- the software one just says "dependency installation". And in principle, some of the setups could separately instruct you to build software instead of assuming you've already done so. I'm concerned that using a shortcode layout would make it harder for someone to easily change the instructions for a specific page if needed.

timothytrippel replied:
Ah got it. SG.

timothytrippel commented on doc/getting_started/install_vivado/index.md:
```diff
@@ -0,0 +1,129 @@
+
+## Device permissions: udev rules
+
+To program FPGAs user using Vivado typically needs to have permission to USB devices connected to the PC.
```
```suggestion
To program FPGAs the user using Vivado typically needs to have permissions to access USB devices connected to the PC.
```

timothytrippel reviewed (approved):
Added a few more suggestions. Modulo, those, LGTM!

jadephilipoom commented:
Okay, updated to address all the feedback so far -- I added it in a new commit, so just reading the diff for the last commit will show all the new updates.

rswarbrick reviewed (approved):
A couple of nitty comments, but this looks like a huge improvement. I'm in favour of merging sooner rather than later and applying any further tweaks as follow-ups.

Thank you: this is awesome!

jadephilipoom commented:
> A couple of nitty comments, but this looks like a huge improvement. I'm in favour of merging sooner rather than later and applying any further tweaks as follow-ups.

Thanks! I had to resolve a merge conflict anyway, so the latest version includes updates after your feedback as well :slightly_smiling_face:

drewmacrae commented:
Thanks for doing this cleanup, in the future we should probably make aliases for the moved pages so the google search index still points to our documentation when we search for it. (I don't know how other people search). I haven't figured out how this would happen but it looks like it's something hugo supports.

jadephilipoom commented:
> Thanks for doing this cleanup, in the future we should probably make aliases for the moved pages so the google search index still points to our documentation when we search for it. (I don't know how other people search). I haven't figured out how this would happen but it looks like it's something hugo supports.
> 
> I only noticed this in retrospect and it will get resolved as the index gets fixed eventually

Good point! Hugo supports it pretty straightforwardly, so I created #11912 to retroactively add them.
